### ISSUE TITLE
Project dependencies may have API risk issues

### ISSUE THREAD

**Issue by PyDeps** [budzianowski/multiwoz#115]
Hi, In **multiwoz**, inappropriate dependency versioning constraints can cause risks.

Below are the dependencies and version constraints that the project is using

```
torch
numpy
nltk==3.4.5
scipy
tqdm
```

The version constraint **==** will introduce the risk of dependency conflicts because the scope of dependencies is too strict.
The version constraint **No Upper Bound** and **\*** will introduce the risk of the missing API Error because the latest version of the dependencies may remove some APIs.

After further analysis, in this project,
The version constraint of dependency **nltk** can be changed to *>=3.2.2,<=3.7*.
The version constraint of dependency **tqdm** can be changed to *>=4.36.0,<=4.64.0*.

The above modification suggestions can reduce the dependency conflicts as much as possible, 
and introduce the latest version as much as possible without calling Error in the projects.

The invocation of the current project includes all the following methods.

<details><summary>The calling methods from the nltk</summary>
 <pre>nltk.util.ngrams
collections.OrderedDict
</pre>
</details>
<details><summary>The calling methods from the tqdm</summary>
 <pre>tqdm.tqdm
</pre>
</details>
<details><summary>The calling methods from the all methods</summary>
 <pre>max_counts.get
numpy.array
self.labels.append
model.model.Model.gen_criterion
range
idx2words.append
values.items
self.policy
utils.util.loadDialogue
re.compile
numpy.set_printoptions
self.Attn.super.__init__
torch.nn.Embedding
torch.no_grad
h_t.repeat.transpose
join.replace
torch.save
os.path.join
model_turns.append
dict.values
slot_name.service_name.join.lower
val.replace
dic.append
dic.extend
domains_in_goal.append
self.loss.item
MultiWozDB
self.policy.state_dict
sys.exit
model.model.Model.train
math.floor
turn.items
valListFile.append
fixDelex.append
input.self.embedding.transpose
open.write
self.output_lang_index2word.has_key
divideData
self.policy.load_state_dict
word_freqs.values
torch.nn.RNN
torch.nn.functional.dropout
random.random
domain.self.dbs.execute
analyze_dialogue
torch.nn.utils.rnn.pad_packed_sequence
tuple
loadDictionaries
self.v.data.normal_
glob.glob
zipfile.ZipFile
self._parseGoal.keys
self.gen_criterion
self.decoder
BeamSearchNode
DecoderRNN
time.time
corpus.extend
val.replace.replace
db_tensor.append
self.v.repeat
nextnodes.append
unicode_to_utf8
open
testListFile.append
Queue.PriorityQueue
booking.append
self.encoder.load_state_dict
list
all
idx.decoder_hiddens.unsqueeze
gen_stats.keys
max
utils.nlp.BLEUScorer.score
torch.nn.functional.dropout.transpose
self.SeqAttnDecoderRNN.super.__init__
max_len.hidden.repeat.transpose
self.decoder.load_state_dict
get_summary_bstate
absl.flags.mark_flags_as_required
l.l.ih_b.data.fill_
self.output_index2word
hyp.split
dic_area.append
numpy.zeros
self.clipGradients
MultiWozEvaluator
sys_turns.append
absl.app.run
self.embedding
whatCellType
BeamSearchNode.eval
os.path.dirname
abs
states_to_correct.items
sql_query.domain.dbs.execute.fetchall
ind.item
torch.cat
self.attn
domain.turn.items
cell_type.strip.strip
model.output_word2index
filter
sqlite3.connect
numpy.append
self.parameters
model.model.Model.loadModel
hh.size
dic_food.append
fin.readlines
self.loss.backward
self.dropout.self.depth.self.cell_type.self.hid_size_enc.self.emb_size.self.input_lang_index2word.len.EncoderRNN.to
argparse.ArgumentTypeError
torch.nn.Linear
torch.optim.Adadelta
self.out
self.EncoderRNN.super.__init__
encoder_outputs.transpose.transpose
SeqAttnDecoderRNN
super
utils.nlp.normalize
min
self.input_lang_word2index.has_key
argparse.ArgumentParser.add_argument
self.getCount
self.DefaultPolicy.super.__init__
unsort_idx.hidden.transpose.transpose
sent.append
encoder_outputs.transpose.size
self.forward
self.optimizer.step
type
l.l.hh_b.data.fill_
tqdm.tqdm
self.rnn
print
self.greedy_decode.append
zip
json.load.items
self.decode
self.encoder.state_dict
usr_turns.append
train
torch.nn.GRU
torch.nn.init.orthogonal_
self._evaluateGeneratedDialogue
self.greedy_decode
EncoderRNN
os.makedirs
os.path.abspath
unsort_idx.outputs.transpose.transpose.contiguous
line.replace
torch.nn.LSTM
idx.data.replace
loadData
i.d.split
model.model.Model.saveModel
self._parseGoal
fixDelex
json.load.keys
self.DecoderRNN.super.__init__
torch.zeros.view
insertSpace
utils.delexicalize.prepareSlotValuesIndependent
t.target_tensor.view
io.open
self.max_len.self.dropout.self.cell_type.self.output_lang_index2word.len.self.hid_size_dec.self.emb_size.SeqAttnDecoderRNN.to
model.input_word2index
utterances.append
ord
nlp.normalize
self.output_lang_word2index.has_key
self.bleuscorer.score
init_gru
corpus_turns.append
torch.bmm
MultiWozEvaluator.evaluateModel
entities.append
bs_tensor.append
dic_price.append
model.gen_criterion.item
unsort_idx.hidden.transpose.transpose.contiguous
sys.strip
target_tensor.size
new_k.log_prob.item
re.sub
unsort_idx.outputs.transpose.transpose
join.split
self.decoder.state_dict
int
utils.nlp.BLEUScorer
createDict
torch.topk
topi.view.squeeze
join.lower
self.hyps.append
nltk.util.ngrams
f.write
model.model.Model
torch.optim.Adam
self.v.size
utils.util.padSequence
unsort_idx.outputs.transpose.transpose.contiguous.transpose
asMinutes
model_corpus.extend
torch.optim.SGD
torch.nn.Parameter
createDelexData
prepareSlotValuesIndependent
numpy.random.seed
torch.manual_seed
utils.dbPointer.oneHotVector
topi.view.detach
self.W_bs
unsort_idx.hidden.transpose.transpose.contiguous.repeat
isinstance
topi.squeeze.detach
self.optimizer.zero_grad
proba.view.view
cPickle.load
input_tensor.size
get_dial
self.W_u
utils.delexicalize.delexicalise
domain.goal.append
model.optimizer.zero_grad
decoder_output.topk
line.strip
input_seqs.transpose.transpose
addBookingPointer.tolist
json.dumps
policy.DefaultPolicy
str.items
os.path.exists
model.model.Model.parameters
torch.nn.functional.log_softmax
trainIters
word_freqs.keys
self.dropout.self.cell_type.self.output_lang_index2word.len.self.hid_size_dec.self.emb_size.DecoderRNN.to
model.model.Model.forward
n.wordid.item
line.strip.split
math.log
sort_idx.embedded.transpose
join.find
domain.dbs.execute
encoder_outputs.data.shape.self.v.repeat.unsqueeze
rnn_input.transpose.transpose
sqlite3.connect.cursor
zip_ref.close
attn_energies.F.softmax.unsqueeze
vars
argparse.ArgumentParser.parse_args
usr.strip
dictionary.items
re.sub.replace
numpy.asarray
numpy.argsort
json.dump
join
encoder_outputs.size.self.v.repeat.unsqueeze
sorted
str.keys
sum
addDBPointer
self.encoder
domains.index
d.items
random.sample
absl.logging.info
collections.Counter.values
utils.util.timeSince
argparse.ArgumentParser
usr.strip.split
dicts.append
key.encode
zip_ref.extractall
collections.Counter.items
self._evaluateRealDialogue
torch.nn.NLLLoss
torch.bmm.transpose
torch.zeros
utterance.append
decoder_output.data.topk
self.input_lang_index2word.has_key
self.bs_size.self.db_size.self.hid_size_enc.self.hid_size_pol.policy.DefaultPolicy.to
float
topi.view.view
get_dial.append
replacements.append
math.fsum
shutil.copy
reduce
absl.flags.DEFINE_string
addBookingPointer
self.Model.super.__init__
set
input_tensor.append
BLEUScorer
topi.detach.view
re.sub.split
clean_states.update
delexicaliseReferenceNumber
numpy.ones
sys.stderr.write
slot.split.replace
new_k.indexes.view
domain.provided_requestables.append
unsort_idx.hidden.transpose.transpose.contiguous.transpose
torch.bmm.squeeze
slot.split
buildDictionaries
torch.nn.utils.clip_grad_norm_
sql_query.domain.self.dbs.execute.fetchall
torch.tanh.unsqueeze
model.optimizer_policy.zero_grad
utils.dbPointer.queryResult
gru.reset_parameters
str
v.lower
operator.itemgetter
ref.split
torch.nn.functional.softmax
idx.encoder_outputs.unsqueeze
self.db.queryResultVenues
math.sqrt
torch.nn.Dropout
Queue.PriorityQueue.get
self.W_db
Queue.PriorityQueue.put
torch.tanh
line.replace.split
format_states
collections.Counter
ent.items
dict
is_ascii
self.setOptimizers
torch.nn.utils.rnn.pack_padded_sequence
UserWarning
math.exp
target_tensor.view
torch.rand
get_slot_name
enumerate
frozenset
filename.strip
self.build_model
model.policy.parameters
var.cuda
nodes.get.append
collections.OrderedDict
format
torch.cuda.is_available
sys.strip.split
target_tensor.append
init_lstm
torch.device
json.load
torch.load
len
self.score
torch.LongTensor
h_t.repeat.repeat
torch.nn.functional.log_softmax.squeeze
torch.tensor
re.findall
main
domain.bstate.keys
re.match
utt.replace.replace
dialogues.items
random.shuffle
unicode
</pre>
</details>

@developer
Could please help me check this issue?
May I pull a request to fix it?
Thank you very much.
